File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode fast = head;
        ListNode prevFast = null;
        ListNode prevSlow = null;
        boolean flag = false;

        while (fast != null) {
            ListNode slow = fast;
            for (int i = 0; i < k; i++) {
                if (fast == null) {
                    flag = true;
                    break;
                }
                prevFast = fast;
                fast = fast.next;
            }
            prevFast.next = null;
            if (prevSlow == null) head = flag ? slow : reverse(slow);
            else prevSlow.next = flag ? slow : reverse(slow);
            prevSlow = slow;
        }

        return head;
    }

    private ListNode reverse(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode newHead = reverse(head.next);
        head.next.next = head;
        head.next = null;

        return newHead;
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* slow = head;
        ListNode* fast = head;

        ListNode dummy(0, head);
        ListNode* prev = &dummy;
        while (fast != nullptr && fast->next != nullptr) {
            for (int i = 0; i < k - 1; i++) {
                if (fast->next != nullptr) {
                    fast = fast->next;
                } else {
                    return dummy.next;
                }
            }

            ListNode* nextGroup = fast->next;
            fast->next = nullptr;
            ListNode* newCurGroupHead = reverseList(slow);
            prev->next = newCurGroupHead;
            slow->next = nextGroup;

            prev = slow;
            fast = slow->next;
            slow = slow->next;
        }

        return dummy.next;
    }

private:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        while (head != nullptr) {
            ListNode* next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};